src/cargo/core/compiler/fingerprint/mod.rs

```diff
@@ -772,10 +774,20 @@ impl LocalFingerprint {
     // TODO: This is allowed at this moment. Should figure out if it makes
     // sense if permitting to read env from the config system.
     #[allow(clippy::disallowed_methods)]
-    fn from_env<K: AsRef<str>>(key: K) -> LocalFingerprint {
+    fn from_env<K: AsRef<str>>(
+        key: K,
+        envs: &BTreeMap<String, Option<OsString>>,
+    ) -> LocalFingerprint {
         let key = key.as_ref();
         let var = key.to_owned();
-        let val = env::var(key).ok();
+        let val = envs
+            .get(key)
+            .and_then(|opt| {
+                opt.as_ref()
+                    .and_then(|os_str| os_str.clone().into_string().ok())
+            })
```

Comment:
https://github.com/rust-lang/cargo/pull/14027#discussion_r1636789147

> iiuc on Windows, `env::var` calls `GetEnvironmentVariableW` which is case insensitive. Since we're recreating our own environment look up, should this matter?

Reply:
https://github.com/rust-lang/cargo/pull/14027#discussion_r1637708208

> I don't have a more in-depth study of windows, I'm very sorry.
> 
> Do you mean that you shouldn't use `env::var` or `env::var_os` to query environment variables? That doesn't seem to work : see: [rerun_if_env_changes](https://github.com/rust-lang/cargo/pull/14027/files#diff-bec1ea82298c9855dced8a66db57ca380f385e70254ca1cd3b8f99152465854dR55) and [rerun_if_env_or_file_changes](https://github.com/rust-lang/cargo/pull/14027/files#diff-bec1ea82298c9855dced8a66db57ca380f385e70254ca1cd3b8f99152465854dR109)

Reply:
@ChrisDenton looking for thoughts on case-insensitive env var lookup on Windows.

When generating a `rerun-if-env-changed` fingerprint, this PR emulates env variable lookup that happens within build scripts by reading the environment we set on build scripts and prioritizing that over `std::env::var` calls (which the build script would implicitly inherit).  The build script and `std::env::var`, iiuc, would be doing case insensitive lookups on Windows.  For the variables we explicitly set on build scripts, we'd be doing a case sensitive lookup.

Reply:
BTW for env snapshot in Cargo we have done this: <https://github.com/rust-lang/cargo/blob/00ea165442eba4bb25d50b754c3a64667110e593/src/cargo/util/context/environment.rs#L132>.

Reply:
> When generating a rerun-if-env-changed fingerprint, this PR emulates env variable lookup that happens within build scripts by reading the environment we set on build scripts and prioritizing that over std::env::var calls (which the build script would implicitly inherit). The build script and std::env::var, iiuc, would be doing case insensitive lookups on Windows. For the variables we explicitly set on build scripts, we'd be doing a case sensitive lookup.

I've not looked at this PR yet but that sounds not ideal. E.g. It's not uncommon for `PATH` to be spelled  `Path` or `path`.

Reply:
If we are already doing case insensitive lookups in other parts of cargo, seems like we should here as well

Reply:
Sorry, I need to confirm this. If it's in a windows environment, getting the environment variables in the build script is not case sensitive?

Reply:
Yes, in the windows environment, look ups are case insensitive.

Reply:
Added `windows` environment case insensitive scenarios and corresponding tests.